recipes-sota/config/files/fuse_actions.sh

```
@@ -0,0 +1,354 @@
+    TOUCH "${REBOOT_SENTINEL_FILE}" 2>/dev/null
+    if  [ "$?" -ne 0 ]; then
+        log "Could not create reboot sentinel file; falling back to normal reboot"
+        # Fall back to using reboot.
+        REBOOT 2>/dev/null
+        if [ "$?" -eq 0 ]; then
+            clear_uboot_vars
+            die "Failed to reboot device"
+        fi
+    fi
+
+    echo '{"status": "need-completion", "message": "rebooting soon"}'
+    return 0
+}
+
+do_complete_install() {
+    if [ -e "${REBOOT_SENTINEL_FILE}" ]; then
+        # This could happen if the reboot path/service is disabled.
+        log "Delaying installation completion due to pending reboot"
+        echo '{"status": "need-completion", "message": "delaying completion due to pending reboot"}'
+        return 0
+    fi
+
+    # Note: if die is called from this point on the die handler will be called.; any failure here
+    # will send a status of 'failed' to aktualizr (see 'on_install_failed()')
+    before_dying 'on_install_failed'
+
+    check_install_vars
+
+    local fuse_status
+    fuse_status=$(FW_PRINTENV fuse_status | SED -Ee 's/\s*fuse_status=(.*)$/\1/')
+
+    case "$fuse_status" in
+        pending|need-reboot)
+            # Probably should never end up here, but in-case we do it should be safe
+            # to attempt another reboot and retry the update.
+            log "Incomplete fuse-status of $fuse_status, attempting another reboot"
+            TOUCH "${REBOOT_SENTINEL_FILE}" 2>/dev/null
+            if  [ "$?" -ne 0 ]; then
+                log "Could not create reboot sentinel file; falling back to normal reboot"
+                # Fall back to using reboot.
+                REBOOT 2>/dev/null
+                if [ "$?" -eq 0 ]; then
+                    clear_uboot_vars
+                    die "Failed to reboot device"
+                fi
+            fi
+            echo '{"status": "need-completion", "message": "delaying completion due to pending reboot"}'
+            ;;
```

For me, these cases look like unexpected failures and as I mentioned before I don't think we should retry the operation.

The "pending" case is likely an indication that the `uEnv.txt` does not have the fusing logic (unsupported machine, for example). We could fail the install operation indicating a possible unsupported operation (as part of the `message` field) in this situation.

As for the "need-reboot" case, I guess it should never happen based on my understanding of the logic in `uEnv.txt`. The only way I see it happening is if the `reset` command in U-Boot did not work which would be a very bad case. In this situation we could fail the install operation indicating an unknown error.

---

Done in V2